### ISSUE TITLE
Transformer: Auto split too long lines

### ISSUE THREAD

**Issue by bhirsz** [MarketSquare/robotframework-tidy#14]
If the line is too long - for example more than 140 chars and of course configurable - it should split argument in multi lines. 

From this (one liner):
```
${var}  ${var2}  My Quite Long And Absurdly Descriptive Keyword That Is Doing Something   ${long_argument1}   ${long_argument2}    ${long_argument3}   ${long_argument4}   ${long_argument5}
```
to
```
${var}  ${var2}  My Quite Long And Absurdly Descriptive Keyword That Is Doing Something
...   ${long_argument1}
...   ${long_argument2}
...   ${long_argument3}
...   ${long_argument4}
...   ${long_argument5}
```
Whether first argument should be on the same line as keyword or not could be configurable. 

We could also add option to always split keywords (well, if someone pass allowable line length to 0) - together with globally passed start/end line it's great way to have quick tool for splitting keywords while writing. Reference: https://github.com/bhirsz/robotframework-clean/blob/fe19cd33a3e1a05845428242bbcfdbd9a5e0cd36/robotclean/code_formatters.py#L10

**Comment by bhirsz** [comment]
I've started working on SplitTooLongLine. It have configurable line_length (default 120) which will split our keyword arguments to multilines if's too long. It also checks already existing multi line keywords if any of the lines does not go over limit (so its not destructive, it leaves alone code if change is not needed. It also preserves original indentation).

It have two possible modes: split every argument to new line (default) or try to fill arguments in one line up to limit (on mobile so example is misaligned) :
```
    Keyword Name ${arg} ${arg} 
   ...     ${arg} ${arg} 
```
 It's also possible to pass 0 to line lengrh limit to ensure that transformer split every line (and you can connect it with global start/end line setting to select part of the file where you want to split all).

Right now I have problem how to handle comments. Given this example:
```
    Keyword  ${arg} # i am comment 
```
If we were to split it it would become:
```
    Keyword
    ... ${arg} 
   ... # i am comment 

```
I was thinking that we could try to leave comment next to keyword name:
```
    Keyword ${arg} # comment
```
to
```
    Keyword # comment
   ... ${arg} 
```
Or move comment before keyword:
```
# comment
    Keyword
    ... {arg} 
```
There are also other options. What do you think? And maybe do you have something it would be good to add to this transformer?

**Comment by MrMino** [comment]
@bhirsz I was hoping to start working on it this Sunday. I have a large body of code that could act as a testbed, as long as I strip the NDA-ed stuff from the outputs. Would love to get notified when you'll put up a WIP PR.

**Comment by MrMino** [comment]
@bhirsz BTW. I think it would be nice to have the comment splitter as a separate entity. You could then make a choice between when you decide to split a comment and how, without bloating the implementation of the argument splitting.

**Comment by bhirsz** [comment]
> @bhirsz I was hoping to start working on it this Sunday. I have a large body of code that could act as a testbed, as long as I strip the NDA-ed stuff from the outputs. Would love to get notified when you'll put up a WIP PR. 

I only have working POC so if you want you can development from me too - and we can discuss in PR if some code could be modified (with comparison how I did it). Whatever you prefer - but it would be nice to see contributions from others.

**Comment by MrMino** [comment]
> I only have working POC so if you want you can take development from me - and we can discuss in PR if some code could be modified (with comparison how I did it).

@bhirsz Sure, let's do that. I can take it as soon as you push the PoC.

> It would be also good occasion to test how robotidy is easy to extend for other people - for example I definietly need to add contribuing guidelines.

Most often I try to infer what's being used based on the code and settings files present in the repository. A contribution guide would surely make that process smother, but it's not indispensable.

Getting a feel for architecture and what-goes-where is a quite different beast. Having a dev guide that outlines that goes a long way in introducing people to a project.

**Comment by bhirsz** [comment]
Pushed code here: https://github.com/MarketSquare/robotframework-tidy/compare/feature/splittoolongline_poc - you can use it as reference only and write it in different way (it's more of quick and dirty code ;) ). Tests will fail because expected files are the same as source (but you can run them to compare how actual and source looks like - it will print diff view but it will also create file under actual/ directory. You can refer to other tests to see how to use multiple sources in tests and compare their results.

Overview of architecture and some general schemes etc are very good idea. I will add them (most likely in next few days).

**Comment by bhirsz** [comment]
Note that we're using RF development version due to recent changes in API that make it easier to use it. It's included in setup.py - but direct download from github can be blocked in your corporate setup (rather through acceptected pip mirrors).

**Comment by mnojek** [comment]
@bhirsz I would opt for moving comment before the keyword but only if the comment is originally in the line where the keyword name is, so I like this suggestion:
```
# comment
Keyword
...    {arg}

**Comment by MrMino** [comment]
@bhirsz I tested your PoC on one of the repos I work with. There were 1139 files changed, so I'll need some time to fully analyze the results. Tomorrow I'll try to run it on an even larger repo. One thing I've noticed so far is that with the [`Run Keyword If`](https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Run%20Keyword%20If), the `ELSE IF` and `ELSE` parts get compressed into a single line, which might not be the desired outcome, as it degrades the readability:

```robot
*** Test Cases ***
Long Run Keyword If Line
    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity}=    Set Variable    123
    Run Keyword If    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity} == 123    Log To Console    This is my log message in a very long line of code
    ...    ELSE    Log to console    Another message
```

```robot
*** Test Cases ***
Long Run Keyword If Line
    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity}=    Set Variable    123
    Run Keyword If    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity} == 123    Log To Console
    ...    This is my log message in a very long line of code    ELSE    Log to console    Another message
```

Should I try to improve upon it in this transformer implementation?

I guess this could be fixed by a pass of a separate transformer ("`PrettifyIfElse`"?). Should such things be kept inside a single transformer or moved to a separate one? I'm still airing on the side of simplifying them as much as possible, but I guess there might be downsides to that approach.

**Comment by bhirsz** [comment]
@MrMino Yes I agree - actually here is my very early prototype (which I used to select text in Pycharm and press shortcuts to split selected keywords into multuline): https://github.com/bhirsz/robotframework-clean/blob/master/robotclean/code_formatters.py#L13 You can see that I also changed behaviour for some of the keywords where I split after condition. ie:
```
*** Test Cases ***
Long Run Keyword If Line
    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity}=    Set Variable    123
    Run Keyword If    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity} == 123    Log To Console    This is my log message in a very long line of code
    ...    ELSE    Log to console    Another message
```
to
```
*** Test Cases ***
Long Run Keyword If Line
    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity}=    Set Variable    123
    Run Keyword If    ${VariableWithALongVerboseNameThatIsBorderingOnPlainInsanity} == 123
    ...    Log To Console    This is my log message in a very long line of code
    ...    ELSE
    ...    Log to console    Another message
```

Adding it as separate transformer is better in my opinion because it's easier to manage what tidy does (and what not). Note that there is 'ReplaceKeywordIf' transformer that will replace all RunKeywordIfs with If's for RF 4.0 so part of the problem will disappear.

**Comment by bhirsz** [comment]
More thoughts:

What to do with too many arguments:
```
Keyword
    [Arguments]    ${argument}    ${argument}    ${argument}    ${argument}    ${argument}    ${argument}    ${argument}
```
Should it be handled by separate transformer SplitTooLongArguments?

What to do with misaligned/too long documentation:
```
[Documentation]    This is first line of this doc and it's going right up to limit. 
....                            this is second line and unfortunately it's going and going and it went outside the limit.
```
Should it be handled by (also) separate transformer ie Split /PrettifyDocs ?

**Comment by pekkaklarck** [comment]
Handling comments when splitting lines is tricky. Notice that
```
    Keyword  ${arg} # i am comment

**Comment by pekkaklarck** [comment]
Splitting arguments, tags, etc. could arguable be covered by the same config the line length. Documentation is harder because you cannot safely split long lines that don't have an explicit separator. Misaligned documentation should be fixed but that should be done by some other transformer.

**Comment by MrMino** [comment]
@pekkaklarck I had doubts whether we want to go this way because of situations like these:

```
Keyword
...  ${arg0}   ${arg1}  ${arg2}  ${...} 
...  ${arg3}   ${arg4}   ${arg5}    # comment about arg5
...  ${some}   ${other}  ${args}
```

It's not immediately obvious what should be done here. If we want the argument splitter to perform comment splits on those examples, where would the comment land? What would be the configurations options? What if there is more than one comment?

**Comment by bhirsz** [comment]
> @pekkaklarck I had doubts whether we want to go this way because of situations like these:
> 
> ```
> Keyword
> ...  ${arg0}   ${arg1}  ${arg2}  ${...} 
> ...  ${arg3}   ${arg4}   ${arg5}    # comment about arg5
> ...  ${some}   ${other}  ${args}
> ```
> 
> It's not immediately obvious what should be done here. If we want the argument splitter to perform comment splits on these, where would the comment land? What would be the configurations options? What if there is more than one comment?

It's tricky. We could also try to keep argument + comment as single entity (split it to new line together) but in this case:
```
Keyword  ${arg}  # i am comment ot whole keyword call
Keyword
...  ${arg}  # i am comment to ${arg}
```
First call would move comment to last line even if it would be better to movei it before comment. Second call would be fine though.

**Comment by MrMino** [comment]
@bhirsz 
So, in essence:

```
Keyword  ${arg}  # i am comment ot whole keyword call

⬇

# i am comment to ${arg}
Keyword
...  ${arg} 
```

but

```
Keyword
...  ${short}   ${short}   ${short}  # long comment about arg5
...  ${long}   ${long}

⬇

Keyword
...  ${short}   ${short}   
...  ${short}  # long comment about arg5
...  ${long}
...  ${long}
```

Hand-wavy rules with multiline keywords in the "pack" mode: 
 - Pack as much args as possible in a single line
 - If an argument has a comment and the comment does not fit, move the arg _and_ the comment to the next line, then go to the next line after that
 - Repeat
 
 Correct?

**Comment by bhirsz** [comment]
@MrMino Yes. This looks least destructive for the code itself. Tip: in ModelTransformer (or rather its parent ast.NodeTransformer) you can return list of nodes instead of node to add to current parent. 
So if you do this:
```
visit_KeywordCall(self, node):
    (...)
    return Comment.from_statement(comment='i am comment to whole keyword call'), node
```
it will create this:
```
    # i am comment to whole keyword call
   <node>
```
Note you need to also pass indent to Comment.from_statements because you could have such situation in for loop / if header and we need to preserve indentation.

**Comment by MrMino** [comment]
Sorry it took so long. Last two weeks were a bit hectic.

I opted not to add the `split_only_at_max_length` as I wasn't sure how to interpret it.

Three things stand out for me so far:

1. If we have a multiline call to split, but the comment is attached before the first argument - should it land before, on, or after the keyword line?

```robot
This is a keyword    # It has its comment here
...  then    there   are   its   args  which  exceed   max length
```

2. Say we have a singleline call without any arguments that has a comment. Suppose it would fit, but only after ensuring proper spacing between the comment and the keyword: should we move the comment above, or leave it attached to the first line?

```robot
This is a keyword                           # suppose this comment does not fit
```

3. Should spacing between kw/args and their comments be the same length as spacing between arguments?

**Comment by bhirsz** [comment]
``split_only_at_max_length`` was the extra feature so it can be added in the future. It might be useful for some (especially if you want to avoid long lines but you don't want keywords spanning for 5-10 lines). What it does is pretty much instead of splitting every argument to new lines we tried to pack as much arguments into one line as possible (up to line length limit). 

1. It looks like comment to keyword itself, it can land before keyword
2. Interesting. There will be whitespace normalizator which should be run before this transformation -> in this case it should change the separator for four spaces and comment would fit. Otherwise if comment is too long it should go before keyword. 
3. Yes, let's go with standard spacing here

I don't have any preferences regarding that matter - we can also ask others and see what's most popular choice. Comment strategy could be added now or in the future.

**Comment by pekkaklarck** [comment]
I think it would be fine in the beginning to move all comments above the split statement. It's easy and works great in most of the cases (especially if you don't have comments at all :smile:). If that is a problem for someone, they can reorganize comments manually when inspecting changes and based on experience this can be then fine-tuned in the future.

**Comment by MrMino** [comment]
@pekkaklarck sure, that's not a showstopper, let's just keep in mind that it's something to think about.

@bhirsz
> What it does is pretty much instead of splitting every argument to new lines we tried to pack as much arguments into one line as possible (up to line length limit).

I thought that's the default of what we're trying to achieve? Couldn't you achieve "split every arg to next line" by setting `line_length=1`?

> Comment strategy could be added now or in the future.

Alright. I'm just making sure that it isn't something I'd have to take into account right now.

**Comment by bhirsz** [comment]
@MrMino 
> I thought that's the default of what we're trying to achieve? Couldn't you achieve "split every arg to next line" by setting `line_length=1`?

My POC splitted 1 argument per 1 line by default (or fit to line length with flag set to True). But I'm fine with being it other way around like you did. Note that if you set line_length to 1 it will always split arguments to new lines. But we could want to `split arguments to new lines (1 arg = 1 line) if they go over line length, otherwise keep it as it is`.

**Comment by pekkaklarck** [comment]
Being able to control are args split to own lines or is as many args as possible fit to each line would be nice and having a flag to control that shouldn't be overly complicated. Even better if there was some heuristic to select the algorithm to use automatically, unless forced otherwise, but that might be something to leave for future.

**Comment by MrMino** [comment]
Alright, I get it now. I'll try to include it after I finish multiline splitting.

**Comment by MrMino** [comment]
If we want the first version to put all comments above the keyword, the implementation gets drastically simpler.

I purposely let myself forget what was the name of the `split_only_at_max_length`, and came up with my own: `split_on_every_arg`. Assuming I understood the purpose of the former correctly, which one should I use?

**Comment by bhirsz** [comment]
If we have 'fill up to max length of line' by default then ``split_on_every_arg`` flag sounds good. And I'm quite bad with names so anything else sounds better :D

**Comment by MrMino** [comment]
Found another interesting edge case (bug?):

```
*** Test Cases ***
Test
    KW
    ...  AAAAAAAAAAAAAA
    ...  # Comment
    ...  BBBBBBBBBBBBBB
```

Robot treats this as if KW received `['AAAAAAAAAAAAAA', '', 'BBBBBBBBBBBBBB']` (notice empty string in the middle) - is this something expected?

Furhtermore, AST seems to think that this:

```
...  # Comment       Comment
```

is composed of the following tokens:
```
Token(CONTINUATION, '...', 8, 4)
Token(ARGUMENT, '', 8, 7)
Token(SEPARATOR, '  ', 8, 7)
Token(COMMENT, '# Comment', 8, 9)
Token(SEPARATOR, '       ', 8, 18)
Token(COMMENT, 'Comment', 8, 25)
Token(EOL, '\n', 8, 32)
```

This complicates matters a great deal, since now a long comment line with two spaces in it would break the tests. The only fix for it is to ensure all moved comments start with a hash, but I have a feeling that it's AST job, not mine :stuck_out_tongue: . @pekkaklarck do you think it is a bug in AST or should I compensate for it inside the transformer?

**Comment by pekkaklarck** [comment]
First of all,

```
Keyword    first
...
...    third
```
being parsed so that `Keyword` gets arguments `['first', '', 'third']` is by design and adding a trailing comments to any of these lines obviously shouldn't change that.

How to handle `# Comment     continued` is a different question. I would be fine changing that so that the result would be thee comment tokens instead of two comments with a separator between. To get that changed for RF 4.0 we needed to act fast. Please ping me on the #devel channel on Slack in that case.

**Comment by MrMino** [comment]
@pekkaklarck I've just noticed that the milestone for 4.0 is on March 11'. I don't think it would be wise to complicate the release few days before the deadline just so that my transformer gets to have one kink less to iron out.

I'll make it work and file an issue to RF tracker, no need to rush it :).
